flocker/control/_persistence.py

```diff
@@ -21,12 +23,98 @@
 from twisted.internet.defer import succeed
 from twisted.internet.task import LoopingCall
 
+from zope.interface import Interface, implementer
+
 from ._model import SERIALIZABLE_CLASSES, Deployment
 
 
 # Serialization marker storing the class name:
 _CLASS_MARKER = u"$__class__$"
 
+# The latest configuration version. Configuration versions are
+# always integers.
+_CURRENT_VERSION = 1
+
+# Formatted bytes representing a versioned config file name.
+_VERSIONED_CONFIG_FILE = b"current_configuration.v%d.json"
+
+# Map of serializable class names to classes
+_CONFIG_CLASS_MAP = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+
+def migrate_configuration(source_version, target_version, config):
+    """
+    Instantiates the applicable configuration migration class for two
+    versions of a persisted configuration and performs a migration.
+
+    :param int source_version: The version to migrate from.
+    :param int target_version: The version to migrate to.
+    :param bytes config: The JSON-encoded source configuration.
+
+    :return bytes: The updated JSON configuration after migration.
+    """
+    migration = (
+        u"_ConfigurationMigration_V%d_V%d"
+        % tuple(sorted([source_version, target_version]))
+    )
+    migration_class = getattr(sys.modules[__name__], migration)
+    config_dict = loads(config)
+    if source_version < target_version:
+        result = migration_class.up(config_dict)
+    else:
+        result = migration_class.down(config_dict)
+    return dumps(result)
+
+
+class _IConfigurationMigration(Interface):
+    """
+    A ConfigurationMigration class provides an interface to migrate between
+    two different versions of a persisted cluster configuration.
+
+    A ConfigurationMigration class must follow a particular naming
+    convention of ``_ConfigurationMigration_Vx_Vy`` where x and y represent
+    the pre and post migration version numbers. This is because the persistence
+    service, upon start, performs automatic sequential upgrades from the most
+    recent configuration format detected on storage to the latest version
+    available and looks for the required migration classes according to
+    this convention.
+
+    Example: a class to migrate between version 1 and 2 configuration formats
+    will be called ``_ConfigurationMigration_V1_V2``.
+    """
+    def up(configuration):
+        """
+        Migrate a Vx source configuration dict to a Vy target
+        configuration format and return the updated dictionary.
+
+        :param dict configuration: The Vx configuration.
+        """
+
+    def down(configuration):
```

Comment:
This seems unnecessary. Downgrading is a significantly harder problem than upgrading, so I'd start with not implementing it.